cirq/experiments/t2_decay_experiment.py

```diff
@@ -163,21 +200,77 @@ def t2_decay(
-            if name not in tab:
-                tab.insert(col_index, name, [0] * tab.shape[0])
+    if num_pulses and len(num_pulses) > 1:
+        max_pulses = max(num_pulses)
+        cols = [f'pulse_{t}' for t in range(max_pulses)]
+        x_basis_measurements['num_pulses'] = x_basis_measurements[cols].sum(
+            axis=1)
+        y_basis_measurements['num_pulses'] = y_basis_measurements[cols].sum(
+            axis=1)
+
+    x_basis_tabulation = _create_tabulation(x_basis_measurements)
+    y_basis_tabulation = _create_tabulation(y_basis_measurements)
 
     # Return the results in a container object
     return T2DecayResult(x_basis_tabulation, y_basis_tabulation)
 
 
+def _create_tabulation(measurements):
+    """Returns a sum of 0 and 1 results per index from a list of measurements.
+    """
+    if 'num_pulses' in measurements.columns:
+        cols = [measurements.delay_ns, measurements.num_pulses]
+    else:
+        cols = [measurements.delay_ns]
+    tabulation = pd.crosstab(cols, measurements.output).reset_index()
+    # If all measurements are 1 or 0, fill in the missing column with all zeros.
+    for col_index, name in [(1, 0), (2, 1)]:
+        if name not in tabulation:
+            tabulation.insert(col_index, name, [0] * tabulation.shape[0])
+    return tabulation
+
+
+def _cpmg_circuit(qubit: devices.GridQubit, delay_var: sympy.Symbol,
+                  max_pulses: int):
+    """Creates a CPMG circuit for a given qubit.
+
+    The circuit will look like:
+      sqrt(X) - wait(delay_var)  - X - wait(2*delay_var) - ... - wait(delay_var)
+    with max_pulses number of X gates.
+
+    The X gates are paramterizd by 'pulse_N' symbols so that pulses can be
+    turned on and off.  This is done to combine circuits with different pulses
+    into the same paramterized circuit.
+    """
+    circuit = circuits.Circuit(
+        ops.Y(qubit)**0.5,
+        ops.WaitGate(value.Duration(nanos=delay_var))(qubit), ops.X(qubit))
+    for n in range(max_pulses):
+        pulse_n_on = sympy.Symbol(f'pulse_{n}')
+        circuit.append(
+            ops.WaitGate(value.Duration(nanos=2 * delay_var *
+                                        pulse_n_on))(qubit))
+        circuit.append(ops.X(qubit)**pulse_n_on)
+    circuit.append(ops.WaitGate(value.Duration(nanos=delay_var))(qubit))
+    return circuit
+
+
+def _cpmg_sweep(num_pulses: List[int]):
+    """Returns a sweep for a circuit created by _cpmg_circuit.
+
+    The circuit in _cpmg_circuit parameterizes the pulses, so this function
+    fills in the parameters for each pulse.  For instance, if we want 3 pulses,
+    pulse_0, pulse_1, and pulse_2 should be 1 and the rest of the pulses should
+    be 0.
+    """
+    pulse_points = []
+    for n in range(max(num_pulses)):
+        pulse_points.append(
+            study.Points(f'pulse_{n}', [1 if p > n else 0 for p in num_pulses]))
+    return study.Zip(*pulse_points)
+
+
 class T2DecayResult:
```

Comment:
My understanding is that this is the trickiest and most subtle piece.  The calibration team is reluctant to even give me a standard way to calculate it, since it is vulnerable to error in several ways.  

I have been avoiding doing it and considering leaving the final calculation to the user unless I can get consensus on how to do this properly.  

For now, out of scope.

Reply:
SGTM